### PR TITLE
D3ASIM-3447: first myco api client release

### PR DESCRIPTION
**Pull request by Hassan754**

### CODE REVIEW

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (faizan2590) [inline src/d3a/models/myco_matcher/external_matcher.py:59]: `validate_authentic_bid_offer_pair ` needs to be adapted as at the moment it validates `bid/offer` based on objects derived from `class Bid` & `class Offer` where as here AFAIS it is received in the form of json_dict

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/external_matcher.py:28]: Yes, will rename it to `publish_offers_bids`

File: src/d3a/models/myco_matcher/__init__.py
Comment (Hassan754) [inline src/d3a/models/myco_matcher/__init__.py:13]: Reminder to revert before merging

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (hannesdiedrich) [inline src/d3a/models/myco_matcher/external_matcher.py:39]: You are not using the `area_uuid` here. I dont see a danger of hash collisions, so the market_ids should be enough to distinguish between the markets, right?

File: src/d3a/models/myco_matcher/__init__.py
Comment (hannesdiedrich) [reply src/d3a/models/myco_matcher/__init__.py:13]: This should be set inside a setup file for an integrationtest.

**Review by hannesdiedrich — COMMENTED** [review]
Left some comments. Will review once all comments on all repos are addressed.

File: src/d3a/d3a_core/simulation.py
Comment (hannesdiedrich) [inline src/d3a/d3a_core/simulation.py:391]: Non-market areas should not publish these events AFAIS.
I propose to move the `publish_event_finish_myco` and `publish_event_tick_myco` and `publish_event_market_cycle_myco ` into the area and check whether the strategy is None. 

Also as a side note, in the future we might have to think about only sending these events for areas that have an active myco client connected to them. But I guess for these initial cases, we can leave it as is.

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (hannesdiedrich) [reply src/d3a/models/myco_matcher/external_matcher.py:37]: Any opinion here @Hassan754 . If you are not using the area_uuid, please use `.values()` instead. Thanks

File: src/d3a/d3a_core/simulation.py
Comment (Hassan754) [reply src/d3a/d3a_core/simulation.py:391]: I'm publishing ticks / markets from the simulation not from areas
Also I contacted Colin and he mentioned that we will only allow 1 myco client to get open offers/bids and post recommendations (and he did not deny that they may leave it open for clients to compete) so he also insisted that I should not put constraints now because they need more time to agree on the process of registration, this is why I;'m only validating if the client is loggedin

Comment (hannesdiedrich) [reply src/d3a/d3a_core/simulation.py:391]: You mixed something up here with my comment on d3a-web.

Here I am just saying that it does not make sense to send the e.g. `publish_event_market_cycle_myco` for a load. Only for a market it makes sense. And This is why i am advocating to move this into the area where you have access to the check `if area.strategy` (if it is None, then it is a market area).

Comment (Hassan754) [reply src/d3a/d3a_core/simulation.py:391]: @hannesdiedrich you're missing the fact the the Myco client is not aware of areas, it is for the whole simulation
pay attention to the difference between it and the api-client , here the client subscribes to the simulation channel whereas in the api-client we used to instantiate objects for each area where it subscribes to this specific area events

Comment (hannesdiedrich) [reply src/d3a/d3a_core/simulation.py:391]: 🤦  I am sorry. You are right of course. If this is settled, then some of my other comments dont make sense as well. Let me re-review. Thanks for bering with me!

File: src/d3a/d3a_core/simulation.py
Comment (Hassan754) [reply src/d3a/d3a_core/simulation.py:364]: Sure, makes sense !

File: src/d3a/d3a_core/simulation.py
Comment (hannesdiedrich) [inline src/d3a/d3a_core/simulation.py:375]: This is getting kind of long and I suggest to refactor this into a function (can be reused in L398 too)

**Review by hannesdiedrich — COMMENTED** [review]
One last refactoring comment. LGTM after addressed.

File: src/d3a/global_utils.py
Comment (hannesdiedrich) [inline src/d3a/global_utils.py:9]: IMO this way of implementing it does not make the check shorter or more readable. You actually only want to check if the match type is custom right? There will not be any other custom match algorithms in the future, so I propose:
```suggestion
def is_custom_matching_enabled():
    return ConstSettings.IAASettings.BID_OFFER_MATCH_TYPE == d3a.constants.BidOfferMatchAlgoEnum.CUSTOM and self.simulation_config.external_connection_enabled
```
I also added the check for the `external_connection_enabled` to make the call in the simulation.py even shorter

**Review by spyrostz — DISMISSED** [review]
LGTM once Hannes' comments are dealt with

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (hannesdiedrich) [reply src/d3a/models/myco_matcher/external_matcher.py:38]: This still needs to be dealt with.

File: src/d3a/global_utils.py
Comment (hannesdiedrich) [inline src/d3a/global_utils.py:1]: Why does this need to be a separate module?

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/external_matcher.py:38]: @hannesdiedrich in the Test 2 we need to pass the area_uuid to the Myco client, this is why I left it here

File: src/d3a/global_utils.py
Comment (Hassan754) [reply src/d3a/global_utils.py:1]: This is to solve some imports issue, it used to reside in a directory 2 levels below this one, and we had to import it from different directories in the same level

File: src/d3a/global_utils.py
Comment (hannesdiedrich) [reply src/d3a/global_utils.py]: What happened to my comment `I also added the check for the external_connection_enabled to make the call in the simulation.py even shorter` ? Didnt you like that?

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (hannesdiedrich) [reply src/d3a/models/myco_matcher/external_matcher.py:38]: OK! Understood.

File: src/d3a/global_utils.py
Comment (hannesdiedrich) [reply src/d3a/global_utils.py:1]: I am not a fan of adding yet another util module. And I really dont understand the import issues. 
Can you try to add it to `d3a/d3a-core/util.py` please?

File: src/d3a/global_utils.py
Comment (Hassan754) [reply src/d3a/global_utils.py]: @hannesdiedrich the `external_connection_enabled` attribute exists in the `self.simulation_config` object which does not exists in the function, we can pass it to it but this won't make the code shorter at all , it's basically the same

Comment (hannesdiedrich) [reply src/d3a/global_utils.py]: (For future PR reviews: This comment would have bee great even before I asked for it. Thanks)

I understand. Then leave it out. Sorry for dragging this cosmetic issue so long. I thought is is a simple change and didnt thing it is such a hassle.

File: src/d3a/d3a_core/rq_job_handler.py
Comment (fievelk) [inline src/d3a/d3a_core/rq_job_handler.py:99]: Where can I find documentation about the `4`?

Comment (fievelk) [reply src/d3a/d3a_core/rq_job_handler.py:99]: Found it! https://github.com/gridsingularity/d3a-web/blob/41bcce49f00dedf83723ac868f0e393ac141236a/simulation/graphql/graphql_enums.py#L28-L31

I really think we should turn these integers into Enums.

File: src/d3a/d3a_core/simulation.py
Comment (fievelk) [inline src/d3a/d3a_core/simulation.py:31]: I would personally not add this, since it just applies to PyCharm.

File: src/d3a/d3a_core/simulation.py
Comment (fievelk) [inline src/d3a/d3a_core/simulation.py:344]: Nit: to follow PEP-8 and avoid using backslashes you could use parentheses:
```suggestion
                if is_custom_matching_enabled():
                    (self.simulation_config.bid_offer_matcher
                        .match_algorithm.publish_market_cycle_myco())
```

File: src/d3a/models/myco_matcher/__init__.py
Comment (fievelk) [inline src/d3a/models/myco_matcher/__init__.py:20]: I would suggest to uniform this and the other checks (here you use a function, while the other `if` clauses don't). Maybe it would even possible to use a dictionary for the matching, like:
```python
MARKET_TO_MATCHING_ALGORITHM_MAPPING = {
    BidOfferMatchAlgoEnum.PAY_AS_CLEAR.value: PayAsClearMatcher,
    BidOfferMatchAlgoEnum.PAY_AS_BID.value: PayAsBidMatcher,
    BidOfferMatchAlgoEnum.CUSTOM.value: ExternalMatcher,
}
```
and then you could just do
```python
self.match_algorithm = MARKET_TO_MATCHING_ALGORITHM_MAPPING[ConstSettings.IAASettings.BID_OFFER_MATCH_TYPE]
```
Or something similar.

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (fievelk) [inline src/d3a/models/myco_matcher/external_matcher.py:11]: To describe this matching strategy, you've always used the `custom` keyword. I think we should uniform the terminology and either use `custom` or `external`, not both.
So I would vote for either changing this class name to `CustomMatcher` (which I'm not so sure about) or changing the `BidOfferMatchAlgoEnum` to `EXTERNAL`. Or a third option, if you have it!

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (fievelk) [inline src/d3a/models/myco_matcher/external_matcher.py:13]:
```suggestion
        super().__init__()
```
This is the python 3 version.

File: src/d3a/models/myco_matcher/__init__.py
Comment (fievelk) [reply src/d3a/models/myco_matcher/__init__.py:20]: If the logic you want to enforce is more complex, you could create a class for it (like `MatchingAlgorithmMapper.get_matching_algorithm()`)

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (fievelk) [inline src/d3a/models/myco_matcher/external_matcher.py:123]: - Nit: I'm not convinced that this docstring says much about the method (especially since it's the same docstring repeated for multiple methods).
- Nit: PEP 257 (the one for docstrings) recommends using a single line when possible. So this would be `"""Myco API"""`. Anyway, this is very minor and I appreciate the fact that you added documentation 👍

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (fievelk) [inline src/d3a/models/myco_matcher/external_matcher.py:31]: Nit, if you're intersted: if you want to follow the PEP recommendations (from [PEP 257](https://www.python.org/dev/peps/pep-0257/#one-line-docstrings)):
> The docstring is a phrase ending in a period. It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...".

So this could become:
```suggestion
        """Query published open offers and bids.
        
        Published data are of the following format: ...
```
(The blank line in between is also part of the PEP for multiline docstrings).

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (fievelk) [inline src/d3a/models/myco_matcher/external_matcher.py:35]: Is this a `TODO` or a regular comment?

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (fievelk) [inline src/d3a/models/myco_matcher/external_matcher.py:51]: Sorry for noticing this late. I think that it would actually be nice to call the channel `offers_bids` (without the `get_`). The reason is that `get` describes only one part of the interactions with the channel, which is used both for publishing and for subscribing. What do you think?

Comment (fievelk) [reply src/d3a/models/myco_matcher/external_matcher.py:51]: Especially for this method, the inconsistency seems evident to me: it is `publish_offers_bids()`, but it's using a channel that is called `get_offers_bids`

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (fievelk) [inline src/d3a/models/myco_matcher/external_matcher.py:67]: Why is this needed?

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (fievelk) [inline src/d3a/models/myco_matcher/external_matcher.py:95]: I would highly recommend replacing assertions with custom exceptions. As a best practice, it's recommended to use assertions only to check for code issues that need to be spotted during the development/debug phase (for coding mistakes, for example). 
For cases in which the program can fail because of external input (like in this case) the best thing is to use exceptions and handle them.

One more reason to use exceptions (that doesn't apply to our current case, but might in the future) is that `assert` statements can be automatically removed if optimizations are requested by Python.

Comment (fievelk) [reply src/d3a/models/myco_matcher/external_matcher.py:95]: Some sources that explain it clearly (in my opinion):
- https://softwareengineering.stackexchange.com/a/225996
- https://stackoverflow.com/questions/40182944/whats-the-difference-between-raise-try-and-assert

Comment (fievelk) [reply src/d3a/models/myco_matcher/external_matcher.py:95]: (So usually you want assertions to fail badly and not to be catched, because they're meant to be seen and fixed by the developer. Exceptions are meant to be catched instead, since they are closely related to the production code.)

**Review by fievelk — COMMENTED** [review]
I left some comments. Some of them are minor (like docstring formatting) while other are more important (like assertions/exceptions and some refactoring). Will review again once the comments are addressed :)

File: src/d3a/models/myco_matcher/__init__.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/__init__.py:20]: This can work for setting the Match Algorithm, however the `is_custom_matching_enabled` is intended as a shortcut for checking if the external matching is enabled, not only for setting the matching algorithm.

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/external_matcher.py:35]: This is a helpful TODO comment for the Test 2

File: src/d3a/models/myco_matcher/external_matcher.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/external_matcher.py:67]: It is a remnant of previous implementation

File: src/d3a/models/myco_matcher/__init__.py
Comment (fievelk) [reply src/d3a/models/myco_matcher/__init__.py]: I'm still not fully convinced. Right now, `is_external_matching_enabled()` has to be called just to decide if we're going to call some specific methods on one specific matcher, and the matcher class itself is instantiated based on this check. So I see some kind of redundancy.
I have the feeling that this could be refactored taking advantage of the standard interface for all matchers (the `BaseMatcher`). We could have one method shared by all children, and they would implement a different internal logic.

File: src/d3a/models/market/__init__.py
Comment (fievelk) [inline src/d3a/models/market/__init__.py:271]: This is a clear improvement over the previous assertions, that's great 👍 
I have another remark to this: it is highly encouraged to avoid raising/catching the base `Exception` class. It would be much better to create a specific exception for this and possibly catch it later on. It could be something generic like `MarketException` or something more specific like `BidOfferValidationException`, or anything really to avoid the base `Exception` :)

Comment (fievelk) [reply src/d3a/models/market/__init__.py:271]: Some more information: https://rules.sonarsource.com/python/RSPEC-112#:~:text=Raising%20instances%20of%20Exception%20and,prone%20and%20difficult%20to%20maintain.

Comment (Hassan754) [reply src/d3a/models/market/__init__.py:271]: The goal of this exception-handling here is to make sure no transaction is posted if any other one will fail, a runtime exception of any type can be thrown here for some other reason besides validation, so we either keep it the way it is now or add new Exception type and add another broad exception handler in the calling code, what do you think ?

File: src/d3a/models/myco_matcher/__init__.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/__init__.py]: The problem that Hannes was pointing to is that in several places in the code we wanted to check if the external matching is enabled [See here](https://github.com/gridsingularity/d3a/blob/ceb0cd569ebd369540e69c5c502f993157bd94fb/src/d3a/d3a_core/simulation.py#L341)
I can encapsulate the algorithm selection in a function if you believe this is an addition, or I can create another function which we can pass the matching type and it verifies if it is set globally, and the `is_external_matching_enabled` will then call it with the corresponding argument.
If you have something else in mind let me know

File: src/d3a/models/market/__init__.py
Comment (fievelk) [reply src/d3a/models/market/__init__.py:271]: > a runtime exception of any type can be thrown here for some other reason besides validation

What kind of exceptions do you think might occur here? In theory, the method does not need to deal with all possible types of exceptions, because other parts of the code should have already dealt with those. For example, you will in general expect `bid.energy_rate` to exist and never raise an `AttributeError`. If you really want to catch all possible types of exceptions (which is not ideal, but it's an option sometimes) you could do something like:
```python
try:
    if not (...):
except Exception as ex:
    raise SomeCustomException(ex) from ex

class SomeCustomException(Exception):
    """Exception raised when this and that happens."""
```
So you would still raise a custom exception, and the other parts of the code will know where it comes from. Also, it will contain the message raised by the original exception, which is necessary to properly debug errors.
At this point, you can either catch it using its name, or with a broad `Exception` (not ideal, but sometimes not easily avoidable).

Comment (fievelk) [reply src/d3a/models/market/__init__.py:271]: I suppose that, to avoid crashes on running simulations, there's a main entry point for d3a where all exceptions are dealt with, catching a broad `Exception`.

File: src/d3a/models/myco_matcher/__init__.py
Comment (fievelk) [reply src/d3a/models/myco_matcher/__init__.py]: My idea is to avoid having more small functions that we need to call multiple times across the code. Also, these functions might be populated with `if` statements, which complicates stuff. That's why I am suggesting alternative ways, more SOLID-friendly, that I believe would be more robust and easier to test.
However, you have the biggest knowledge of this code. So if you believe that my suggestion is not practical and the current implementation is a better way to proceed, I will give my LGTM 👍

Comment (Hassan754) [reply src/d3a/models/myco_matcher/__init__.py]: I will keep working in the same area in Test 2, will try to tackle this then

File: src/d3a/d3a_core/redis_connections/aggregator_connection.py
Comment (spyrostz) [inline src/d3a/d3a_core/redis_connections/aggregator_connection.py:70]: I assume the lazy import is required in order to be able to instantiate the Redis connection only if needed, right ?

Comment (Hassan754) [reply src/d3a/d3a_core/redis_connections/aggregator_connection.py:70]: Not really, there is here to solve a  circular import issue
I had the map of the imports on paper, I can draw you the whole setup if needed

Comment (fievelk) [reply src/d3a/d3a_core/redis_connections/aggregator_connection.py:70]: Nit: do you think there are other ways to solve this? Circular imports often point out some issues with the package structure so it can be good to deal with it. Not blocking though.

File: src/d3a/models/myco_matcher/__init__.py
Comment (fievelk) [inline src/d3a/models/myco_matcher/__init__.py:12]: What was the reason to make this a normal method, rather than using `__init__` as before? I think it would be good to document it in the docstring.

File: src/d3a/models/myco_matcher/__init__.py
Comment (fievelk) [inline src/d3a/models/myco_matcher/__init__.py:22]: Nit:
```suggestion
            raise WrongMarketTypeException("Wrong market type setting flag "
```

File: src/d3a/d3a_core/redis_connections/aggregator_connection.py
Comment (Hassan754) [reply src/d3a/d3a_core/redis_connections/aggregator_connection.py:70]: Yes sure I prefer restructuring the modules, maybe in the context of another ticket

File: src/d3a/models/myco_matcher/__init__.py
Comment (Hassan754) [reply src/d3a/models/myco_matcher/__init__.py:12]: This is because we need the MycoMatcher singleton at startup, but the matcher should be set later on

File: src/d3a/models/area/__init__.py
Comment (fievelk) [inline src/d3a/models/area/__init__.py:396]: Could you please explain what happens here?
I'm not very convinced about this part: it interacts a bit too deeply with the internals of the matcher (without using any interface method). This makes testing trickier, among other stuff. I think it would be cool if we could refactor this somehow, but not necessarily in the context of this PR. Maybe for the next test, if that's ok with you.

Comment (Hassan754) [reply src/d3a/models/area/__init__.py:396]: This is updating the mapping of areas: open offers / bids
I agree with you that this should be using an interface, I'm touching this part in my current test to only update if the Area is not a device, I'll also add the interface there too
Thanks

File: integration_tests/steps/two_sided_pay_as_clear.py
Comment (fievelk) [inline integration_tests/steps/two_sided_pay_as_clear.py:110]: Nit:
```suggestion
               bid_offer_matcher.match_algorithm.state.clearing.values())
```

Comment (fievelk) [reply integration_tests/steps/two_sided_pay_as_clear.py:110]: Same applies below

**Review by fievelk — DISMISSED** [review]
LGTM. 
I would like to have some more documentation in the singleton `init()` method: singletons are quite delicate beasts, so it would be great to explicitly document why we're using them and how.
Among other things, I like the refactoring of Offers/Bids as dataclasses, I think that's a cool improvement for the future.

**Review by spyrostz — APPROVED** [review]
One clarification comment, LGTM though

File: src/d3a/d3a_core/redis_connections/aggregator_connection.py
Comment (spyrostz) [reply src/d3a/d3a_core/redis_connections/aggregator_connection.py:70]: I also agree with Pierpaolo, that in the long run we should restructure the modules to avoid the circular import, not a blocker for now though. In general, we have the implicit rule to only allow lazy imports in the d3a-web integration tests (e.g. in order to avoid imports from models.py before the test suite started Django), we are not too strict on this rule though therefore it is okay.

**Review by spyrostz — APPROVED** [review]
LGTM

**Review by fievelk — APPROVED** [review]

**Pull request merged**
